packages/wallet-ui/src/hooks/useHasMetamask.ts

```diff
@@ -1,22 +1,96 @@
+  request(options: { method: string }): Promise<void>;
+}
+
+declare global {
+  interface Window {
+    ethereum?: MetaMaskProvider;
+  }
+}
+
+function isMetaMaskProvider(obj: unknown): obj is MetaMaskProvider {
+  return obj !== null && typeof obj === 'object' && obj.hasOwnProperty('isMetaMask') && obj.hasOwnProperty('request');
+}
+
+function detectMetaMaskProvider(
+  windowObject: Window & typeof globalThis,
+  { timeout = 3000 } = {},
+): Promise<MetaMaskProvider | null> {
+  let handled = false;
+  return new Promise<MetaMaskProvider | null>((resolve) => {
+    const handleEIP6963Provider = (event: CustomEvent) => {
+      const { info, provider } = event.detail;
+      if (['io.metamask', 'io.metamask.flask'].includes(info.rdns) && isMetaMaskProvider(provider)) {
+        resolve(provider);
+        handled = true;
+      }
+    };
+
+    if (typeof windowObject.addEventListener === 'function') {
+      windowObject.addEventListener('eip6963:announceProvider', (event: Event) => {
+        handleEIP6963Provider(event as CustomEvent);
+      });
+    }
+
+    setTimeout(() => {
+      if (!handled) {
+        resolve(null);
+      }
+    }, timeout);
+
+    // Notify event listeners and other parts of the dapp that a provider is requested.
+    if (typeof windowObject.dispatchEvent === 'function') {
+      windowObject.dispatchEvent(new Event('eip6963:requestProvider'));
+    }
+  });
+}
+
+async function waitForMetaMaskProvider(
+  windowObject: Window & typeof globalThis,
+  { timeout = 1000, retries = 0 } = {},
+): Promise<MetaMaskProvider | null> {
+  return detectMetaMaskProvider(windowObject, { timeout })
+    .catch(function () {
+      return null;
+    })
+    .then(function (provider) {
+      if (provider || retries === 0) {
+        return provider;
+      }
+      return waitForMetaMaskProvider(windowObject, {
+        timeout,
+        retries: retries - 1,
+      });
+    });
+}
+
+async function detectMetamaskSupport(windowObject: Window & typeof globalThis) {
+  const provider = await waitForMetaMaskProvider(windowObject, { retries: 3 });
+  return provider;
+}
+
 export const useHasMetamask = () => {
   const dispatch = useAppDispatch();
   const [hasMetamask, setHasMetamask] = useState<boolean | null>(null);
+
   useEffect(() => {
     const init = async () => {
       try {
         dispatch(enableLoadingWithMessage('Detecting Metamask...'));
-        //make sure mm has installed
-        if (await detectMetamask()) {
-          //metamask SDK is not support when multiple wallet installed, and each wallet may injected window.ethereum, some may override isMetamask
-          const _provider = await getProvider();
-          dispatch(setProvider(_provider));
-          setHasMetamask(_provider != null);
+        const provider = await detectMetamaskSupport(window);
+        // Use the new detection method
+
+        if (provider && (await isSupportSnap(provider))) {
+          window.ethereum = provider;
```

Comment:
i guess we dont need to assign window.ethereum?
and the return should be just the provider